### ISSUE TITLE
Syntax errors when using union with first/limit in 1.0.4

### ISSUE THREAD

**Issue by fnatte** [knex/knex#5061]
# Environment

Knex version: 1.0.4
Database + version: PostgreSQL 13.4
OS: Debian

# Bug

In 1.0.4 the behavior (order and wrapping) of union changed causing queries that previously worked to generate syntax errors. 

One example is:
```ts
const results = await knex('accounts')
  .select('last_name')
  .unionAll(function () {
    this.select('last_name').from('accounts');
  })
  .first();
expect(results).to.eql({ last_name: 'User' });
```

Which in 1.0.4 gives `error: select "last_name" from "accounts" limit $1 union all select "last_name" from "accounts" - syntax error at or near "union"`.

But in 1.0.3 it generates `select "last_name" from "accounts" union all select "last_name" from "accounts" limit ?`.

Another query that fails is
```ts
const results = await knex
  .unionAll(
    function () {
      this.select('last_name').from('accounts');
    },
    function () {
      this.select('last_name').from('accounts');
    },
    true
  )
  .first();
expect(results).to.eql({ last_name: 'User' });
```

with `error: (select "last_name" from "accounts") union all (select "last_name" from "accounts") (limit $1) - syntax error at or near "("`

I believe this change is introduced in https://github.com/knex/knex/pull/5030. Is this new behavior intended and we should rewrite our union queries?

This following is working, so it's not that hard to update the queries but it seems like a breaking change?
```ts
const results = await knex
  .unionAll(
    function () {
      this.select('last_name').from('accounts');
    },
    function () {
      this.select('last_name').from('accounts');
    }
  )
  .first();
```

**Comment by kibertoad** [comment]
@OlivierCavadenti

**Comment by wa11a** [comment]
we got the same issue with postgres after upgrade. the problem we encountered is that we had a limit anf knex placed the limit inside the first statement in the union. and union statements can't have limit. only the top level query can have a limit.
downgraded to 1.0.3

**Comment by KristjanTammekivi** [comment]
Looks like I just hit this with .orderBy(...) as well

**Comment by OlivierCavadenti** [comment]
I will take more time to analyze this because it seems that the two views of how unions are constructed are wanted:
- put limit, orderBy, ... in main query when needed, and wrap them : https://github.com/knex/knex/issues/913 https://github.com/knex/knex/issues/913#issuecomment-334481807
- put limit at the end of the query.

Maybe I will think about define default position of clauses with unions and add options to change orders/wrapping.

**Comment by felixmosh** [comment]
Hit the same issue, using `Order By` on union query.

```js
await db
  .unionAll(
    [
      db('customers')
        .column('phone', db.raw(`? as global_idx`, [1]))
        .where('customers.id', 1),
      db('accommodations')
        .column('displayPhone', db.raw(`? as global_idx`, [2]))
        .where('accommodations.id', 2),
    ],
    true
  )
  .orderBy(['global_idx']);
```

Outputs:
```sql
// knex 1.0.3

(select `phone`, 1 as global_idx from `customers` where `customers`.`id` = 1) union all (select `display_phone`, 2 as global_idx from `accommodations` where `accommodations`.`id` = 2) order by `global_idx` asc
```

```sql
// knex 1.0.4
(select `phone`, 1 as global_idx from `customers` where `customers`.`id` = 1) union all (select `display_phone`, 2 as global_idx from `accommodations` where `accommodations`.`id` = 2) (order by `global_idx` asc)
```

**Comment by dimip1606** [comment]
Same as @felixmosh, but using `limit` instead of `orderBy` after the `unionAll`.

**Comment by OlivierCavadenti** [comment]
Sorry for the wait, and thanks for all examples guys.
I have find how to fix and I added a lot of unit tests but still need more tests.
I will push tomorrow at the end of day.

**Comment by kibertoad** [comment]
Released in 1.0.5
